### PR TITLE
Rayleigh Gauss Newton

### PR DESCRIPTION
**Pull request by chrisrothUT**
Here's a draft of the PR to implement the RGN method as done in https://arxiv.org/pdf/2106.10558.pdf. I've only implemented the holomorphic case and it appears to be working on real-valued neural networks. For now RGN is a separate driver. Here's how to use it:

```
rgn = nk.optimizer.RGN(solver=cg, diag_shift=1e-3)
vstate = nk.vqs.MCState(sampler=sa, model=ma, n_samples=1000,n_discard_per_chain=9)
sched = np.full([50],0.2)
gs = nk.driver.VMC_RGN(ha, sched, variational_state=vstate,preconditioner=rgn)
gs.run(n_iter=50,out='out')
```

Since the learning rate is built into the preconditioner, for now you just specify an array of learning rates at different time steps. The paper above says you can increase the LR exponentially as you go along. A few orders of business:

1. RGN really shouldn't be a separate driver but it needs access to the average energy and a learning rate parameter, epsilon. So we'd have to amend this line in VMC: `self._dp = self.preconditioner(self.state, self._loss_grad)` if we wanted to use the same driver.

2. I'm not sure about the nuances for non-holomorphic complex valued functions. Perhaps @attila-i-szabo can help?

3. The documentation is all wrong from copy-pasting code. I'll fix that in the coming days

### CODE REVIEW

**Comment by attila-i-szabo** [comment]
Good stuff, thanks for doing this! I'll have a closer look at it in the coming days, here are some high-level comments in the meanwhile:

> Since the learning rate is built into the preconditioner, for now you just specify an array of learning rates at different time steps. The paper above says you can increase the LR exponentially as you go along.

This should be a function that maps the number of the time step to `eps` (maybe with support for keeping it constant). This is how optimiser schedules are defined in OPTAX, so it would be good to be consistent with it. Either way, this is internal to the preconditioner, so we should make the schedule of `eps` a parameter of `RGN` and make `VMC` pass the step count as a parameter of `self.preconditioner` too.

> 1. RGN really shouldn't be a separate driver but it needs access to the average energy and a learning rate parameter, epsilon. So we'd have to amend this line in VMC: `self._dp = self.preconditioner(self.state, self._loss_grad)` if we wanted to use the same driver.

I support this, with the same caveat as above. SR precodnitioners are easy to modify, you just have to make them take and ignore these arguments.

> 2. I'm not sure about the nuances for non-holomorphic complex valued functions. Perhaps @attila-i-szabo can help?

I'll try to have a look at this at some point. As for non-holomorphic "complex" functions, the `PyTree` implementation can only support them if we split the real and imaginary parts of the parameters explicitly. Then, if everything works for real parameters, these should be fine too.

More importantly, eq. (11) should be checked carefully for both real and holomorphic parameters. Much of the literature on RGN (like [this](https://journals.aps.org/prl/pdf/10.1103/PhysRevLett.94.150201) or [this](https://arxiv.org/pdf/cond-mat/0502553.pdf)) is extremely cavalier with complex differentiation or just uses real functions of real parameters everywhere.

---

* Just checking: the `chunk_size` in the RGN constructor behaves rather differently from that in the loss function calculation, right? I mean, a chunk of 1 is a sample and all its neighbours, not just a single NN evaluation. (This would be an argument for using gradients of an appropriately `stop_gradient`ed loss function directly, so we don't have to worry about this.)
* We should look into an `OnTheFly` evaluation. It would probably be super slow, but that's our only hope of pushing this algorithm to a useful number of samples.

**Comment by attila-i-szabo** [comment]
To me, the most pressing issue with RGN is making it performant, i.e., not having to redo the local energy calculation several times. This might not be possible with the `OnTheFly` strategy, but we should think about it.

As for the precomputing strategy, the first term of eq. (11) (the computationally expensive part) includes the Jacobian of the local energy, provided the denominators are `stop_gradient`ed. This might allow us to compute the local energy for `expect_and_grad` and these Jacobians in one go, by putting the local energy function through `jax.jvp`. There are a couple of caveats:
* This is only efficient if we do the gradients of each sample separately (cf. `QGTOnTheFly`), which might be complicated with chunking.
* It would require integrating the preconditioner, the vstate, and the VMC driver much more closely than they are now. We don't want to do this sort of calculation for SR, so `expect_and_grad` should be aware that we will need the energy Jacobian too. I don't know how we can do this without object orgy.

The one upshot for the `OnTheFly` strategy (provided we implement it) is that the energy JVP in the first term should be implemented through `jax.jvp` on a `stop_gradient`ed local energy function, which would be both more readable and more robust than calculating the same gradient by hand.

File: netket/driver/vmc_rgn.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    PreconditionerT,
+)
+from netket.utils import warn_deprecation
+from netket.optimizer import Sgd
+
+from .vmc_common import info
+from .abstract_variational_driver import AbstractVariationalDriver
+
+
+class VMC_RGN(AbstractVariationalDriver):
```
Comment (PhilipVinc) [inline netket/driver/vmc_rgn.py:35]: If you make this a subclass of VMC then you don't need to redefine everything, you just need to define the new `forward_and_backward`

```suggestion
class VMC_RGN(VMC):
```

File: netket/driver/vmc_rgn.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+            optimizer: Determines how optimization steps are performed given the
+                bare energy gradient.
+            preconditioner: Determines which preconditioner to use for the loss gradient.
+                This must be a tuple of `(object, solver)` as documented in the section
+                `preconditioners` in the documentation. The standard preconditioner
+                included with NetKet is Stochastic Reconfiguration. By default, no
+                preconditioner is used and the bare gradient is passed to the optimizer.
+        """
+
+        if variational_state is None:
```
Comment (PhilipVinc) [inline netket/driver/vmc_rgn.py:66]: Assuming you change the inheritance above, you can replace this whole method with
```suggestion
        super().__init__(*args, **kwargs)
        self.schedule = schedule
```

**Review by PhilipVinc — COMMENTED** [review]
If you want me to give a look, can you cleanup unused logic in `rgn_pytree` first? running `flakehell lint netket/optimizer/rgn` can help to do it

File: netket/driver/vmc_rgn.py (hunk 1 above)
Comment (chrisrothUT) [reply netket/driver/vmc_rgn.py:35]: Yes this will definitely simplify things

**Comment by chrisrothUT** [comment]
> This might allow us to compute the local energy for expect_and_grad and these Jacobians in one go, by putting the local energy function through jax.vjp. 

Okay this is a bit more complicated. The issue is the Jacobian needs the gradients of the activations i.e. f'(h) whereas the forward pass needs f(h). Unless jax has a prebuilt function (forward and jacobian) or something like that, this looks very cumbersome to implement

**Comment by attila-i-szabo** [comment]
> I agree with this. It would be nice if you could just pass an OPTAX object but as it turns out they are hard to take apart as they don't store information about the parameters of the optimizer.

Yeah, OPTAX is not really prepared to tune arbitrary hyperparameters, but that's OK, I'm only talking about the interface.
If we are to keep RGN a separate driver, it might be better to leave the schedule as a part of that, on second thoughts it doesn't really fit the idea of the `LinearOperator`.

> Since we're using a different driver anyway, we may as well just call something like `expect_and_grad_and_localenergyjacobian`

and make this function part of the driver rather than `MCState`.

> Okay this is a bit more complicated. The issue is the Jacobian needs the gradients of the activations i.e. f'(h) whereas the forward pass needs f(h). Unless jax has a prebuilt function (forward and jacobian) or something like that, this looks very cumbersome to implement

`jax.vjp` returns the function output as well as the `vjpfun`. I would do something like this:
```python
def energy_and_localenergyjacobian(fun, params, σ, σp, mels):
    def Eloc_and_jacobian(params, σ1, σp1, mels1):
        def Eloc(p):
            return jnp.sum(mels1 * jnp.exp(fun(p, σp1) - jax.lax.stop_gradient(fun(p, σ1))))
        el, vjpfun = jax.vjp(Eloc, params)
        return el, vjpfun(1.0)
    return jax.vmap(Eloc_and_jacobian, in_axes=..., out_axes=...)(params, σ, σp, mels)
```
I'm sure much of this is already implemented here and there, with chunking, this is more for illustration.
I've written this with the `stop_gradient`ed local energy I talked about earlier. The reason (other than elegance) is that atm we do the forward pass _three_ times, once in `expect_and_grad`, once when we redo the forward pass in the Hessian estimation, and once when we backpropagate in the same function. I'm sure the last one can be eliminated in the current implementation if we take care, but his might be simpler at that point.

> An OnTheFly evaluation is going to take a factor of mean(neighbors) longer than a QGTOnTheFly evaluation which is already a bottleneck. Personally, I think the best way to scale up RGN and SR is to make them MPI friendly. The idea would be to have each computer calculate and store the preconditioner for some of the samples and do the `mat_vec` in parallel. We might have to implement a solver by hand as well

Fair enough, it might still be useful to have an implementation around, but it's not so pressing. I think SR is already MPI friendly, @PhilipVinc? Or at least I don't think we have any MPI issues with out-of-the box JAX solvers

File: netket/optimizer/rgn/rgn_pytree_logic.py [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+        params : a pytree of parameters p
+        samples : an array of n samples σ
+
+    Returns:
+        The Jacobian matrix ∂/∂pₖ ln Ψ(σⱼ) as a PyTree
+    """
+
+    def _rhessian_real_holo(logpsi, params, σ, σp, mels):
+        y, vjp_fun = jax.vjp(logpsi, params, σp)
+        scale = mels*jnp.exp(logpsi(params,σp) - logpsi(params,σ))
```
Comment (attila-i-szabo) [inline netket/optimizer/rgn/rgn_pytree_logic.py:133]:
```suggestion
        scale = mels*jnp.exp(y - logpsi(params,σ))
```
This is exactly what I meant.

File: netket/optimizer/rgn/rgn_pytree_logic.py [hunk 4]
```diff
@@ -0,0 +1,358 @@
...
+
+def mat_vec(v: PyTree, oks: PyTree, rhes: PyTree, mean_grad: PyTree, eps: float, en: float, diag_shift: float) -> PyTree:
+    """
+    Compute ⟨O† O⟩v = ∑ₗ ⟨Oₖᴴ Oₗ⟩ vₗ
+    """
+    res = tree_conj(_vjp(oks, _jvp(rhes, v).conjugate()))
+    res2 = tree_conj(jax.tree_map(lambda x: _jvp(mean_grad, v)*x,v))
+    res3 = tree_conj(_vjp(oks, _jvp(oks, v).conjugate()))
+
+    res = jax.tree_multimap(lambda w,x,y,z: w - x - en*y + (1/eps)*(y + diag_shift*z),res,res2,res3,v)
```
Comment (attila-i-szabo) [inline netket/optimizer/rgn/rgn_pytree_logic.py:354]: This does more matrix-vector products than needed. Also, the output vector is typically a lot smaller than the parameter vector, so we should rather sum what we can at that stage.
```suggestion
    # terms involving E_loc,j and the QGT
    w = _jvp(rhes, v) + (1/eps - en) * _jvp(oks, v)
    res = tree_conj(_vjp(oks, w.conjugate()))
    # g_i <O_j> term
    res2 = tree_conj(jax.tree_map(lambda x: _jvp(mean_grad, v)*x,v))

    res = jax.tree_multimap(lambda x,y,z: x - y + (diag_shift/eps) * z,res,res2,v)
```

**Review by attila-i-szabo — CHANGES_REQUESTED** [review]

File: netket/optimizer/rgn/rgn_pytree_logic.py (hunk 3 above)
Comment (attila-i-szabo) [reply netket/optimizer/rgn/rgn_pytree_logic.py:133]: It just needs to be integrated with `expect_and_grad` and so probably moved to the driver

**Comment by PhilipVinc** [comment]
>Fair enough, it might still be useful to have an implementation around, but it's not so pressing. I think SR is already MPI friendly, @PhilipVinc? Or at least I don't think we have any MPI issues with out-of-the box JAX solvers

Solvers only perform matrix-vector products. 
Matrix vector products of QGTs correctly perform reductions over MPI so... 
solvers don't need to know about MPI if the objects they operate upon work correctly with MPI.
QGT do. You'd want your HessianOperator to do the same. in the matvec function

**Comment by attila-i-szabo** [comment]
@chrisrothUT could you clean up this branch? You shouldn't merge `master` into your branch  but rebase on top of it, because now it is full of changes that don't belong here and would complicate merging it eventually.

**Comment by chrisrothUT** [comment]
> @chrisrothUT could you clean up this branch? You shouldn't merge master into your branch but rebase on top of it, because now it is full of changes that don't belong here and would complicate merging it eventually.

Yes I will do that at some point
